code/xcvm/cosmwasm/contracts/gateway/src/interpreter.rs

```diff
@@ -0,0 +1,114 @@
+use crate::{
+	contract::INSTANTIATE_INTERPRETER_REPLY_ID,
+	error::{ContractError, Result},
+	events::make_event,
+	network::load_this,
+	state::{self, interpreter},
+};
+use cosmwasm_std::{
+	to_binary, Deps, DepsMut, Reply, Response, StdError, StdResult, Storage, SubMsg, WasmMsg,
+};
+use cw_xc_interpreter::contract::{
+	XCVM_INTERPRETER_EVENT_DATA_ORIGIN, XCVM_INTERPRETER_EVENT_PREFIX,
+};
+use xc_core::{CallOrigin, InterpreterOrigin, NetworkId};
+
+use crate::{auth, prelude::*};
+
+pub(crate) fn force_instantiate(
+	auth: auth::Admin,
+	gateway: Addr,
+	deps: DepsMut,
+	user_origin: Addr,
+) -> Result {
+	let config = load_this(deps.storage)?;
+	let interpreter_code_id = match config.gateway.expect("expected setup") {
+		GatewayId::CosmWasm { interpreter_code_id, .. } => interpreter_code_id,
+	};
+
+	let call_origin = CallOrigin::Local { user: user_origin };
+	let interpreter_origin = InterpreterOrigin {
+		user_origin: call_origin.user(config.network_id),
+		salt: b"default".to_vec(),
+	};
+	let msg = instantiate(deps.as_ref(), gateway, interpreter_code_id, &interpreter_origin)?;
+	Ok(Response::new().add_submessage(msg).add_event(make_event("interpreter.forced")))
+}
+
+pub fn instantiate(
+	deps: Deps,
+	admin: Addr,
+	interpreter_code_id: u64,
+	interpreter_origin: &InterpreterOrigin,
+) -> Result<SubMsg, ContractError> {
+	let next_interpreter_id: u128 = state::interpreter::INTERPRETERS_COUNT.load(deps.storage)? + 1;
+	let instantiate_msg = WasmMsg::Instantiate2 {
+		admin: Some(admin.clone().into_string()),
+		code_id: interpreter_code_id,
+		msg: to_binary(&cw_xc_interpreter::msg::InstantiateMsg {
+			gateway_address: admin.clone().into_string(),
+			interpreter_origin: interpreter_origin.clone(),
+		})?,
+		funds: vec![],
+		label: ["xcvm_interpreter", &next_interpreter_id.to_string()].join("_"),
```

Comment:
Since you’re using `to_string` anyway, this is probably better as format:

```suggestion
		label: format!("xcvm_interpreter_{}", next_interpreter_id),
```

Reply:
cw is allocating strings like a hell overall, it is too annoying to think when or when not and do such fixes, as long as cw is json, it is premature optimization to optimize format.

Reply:
That’s true but in this instance format is also shorter and more readable.